src/dataframe/dataframe.jl

```diff
@@ -625,58 +624,122 @@ julia> insertcols!(d, 1, b=['a', 'b', 'c'])
+                     makeunique::Bool=false, copycols::Bool=true)
     if !(0 < col_ind <= ncol(df) + 1)
-        throw(ArgumentError("attempt to insert a column to a data frame with $(ncol(df)) columns at index $col_ind"))
+        throw(ArgumentError("attempt to insert a column to a data frame with " *
+                            "$(ncol(df)) columns at index $col_ind"))
     end
-    if !(size(df, 1) == length(item) || size(df, 2) == 0)
-        throw(DimensionMismatch("length of new column ($(length(item))) must match the number of rows in data frame ($(nrow(df)))"))
+
+    if !makeunique
+        if !allunique(first.(name_cols))
+            throw(ArgumentError("Names of columns to be inserted into a data frame must be " *
+                                "unique when `makeunique=true`"))
+        end
+        for (n, _) in name_cols
+            if hasproperty(df, n)
+                throw(ArgumentError("Column $n is already present in the data frame " *
+                                "which is not allowed when `makeunique=true`"))
+            end
+        end
     end
 
-    if hasproperty(df, name)
-        if makeunique
-            k = 1
-            while true
-                # we only make sure that new column name is unique
-                # if df originally had duplicates in names we do not fix it
-                nn = Symbol("$(name)_$k")
-                if !hasproperty(df, nn)
-                    name = nn
-                    break
-                end
-                k += 1
+    if ncol(df) == 0
+        target_row_count = -1
+    else
+        target_row_count = nrow(df)
+    end
+
+    for (n, v) in name_cols
+        if v isa AbstractVector
+            if target_row_count == -1
+                target_row_count = length(v)
+            elseif length(v) != target_row_count
+                throw(DimensionMismatch("length of new column $n which is $(length(v)) must match" *
+                                        " the target number of rows in data frame ($target_row_count)"))
             end
-        else
-            msg = "Duplicate variable name $name. Pass makeunique=true" *
-                  " to make it unique using a suffix automatically."
-            throw(ArgumentError(msg))
+        elseif v isa AbstractArray
+            throw(ArgumentError("AbstractArray other than AbstractVector is not allowed " *
+                                "to be added as a column of a data frame."))
         end
     end
-    insert!(index(df), col_ind, name)
-    insert!(_columns(df), col_ind, item)
-    df
-end
+    if target_row_count == -1
+        target_row_count = 1
+    end
 
-insertcols!(df::DataFrame, col_ind::Int, name_col::Pair{Symbol}; makeunique::Bool=false) =
-    insertcols!(df, col_ind, name_col[1] => upgrade_scalar(df, name_col[2]), makeunique=makeunique)
+    for (name, item) in name_cols
+        if !(item isa AbstractVector)
+            if iterm isa AbstractArray
```

Comment:
```suggestion
            if item isa AbstractArray
```
But maybe turn this into `@assert`

Reply:
OK - I will make an @assert